onnxruntime/core/session/inference_session.cc

```diff
@@ -1733,59 +1732,87 @@ common::Status InferenceSession::Run(IOBinding& io_binding) {
 }
 
 #ifdef ENABLE_TRAINING
-common::Status InferenceSession::RunInBackgroundAndWaitForYield(RunOptions& run_options, IOBinding& io_binding,
-                                                                std::vector<OrtValue>& user_outputs) {
-  const int64_t main_thread_event_id = 0;
-  onnxruntime::contrib::OrtEventPool::GetInstance().ResetEvent(0);
+common::Status InferenceSession::RunInBackgroundAndWaitForYield(const RunOptions& run_options, IOBinding& io_binding,
+                                                                std::vector<OrtValue>& user_outputs, int64_t& run_id) {
+  std::promise<void> setup_promise;
+  std::future<void> setup_future = setup_promise.get_future();
+
+  auto bg_thread = std::thread([&](std::future<void> setup_future) {
+    // wait until task is properly setup
+    setup_future.get();
 
-  task_.terminate_flag_ = &(run_options.terminate);
-  task_.bg_thread_promise_ = std::promise<Status>();
-  task_.bg_thread_future_ = task_.bg_thread_promise_.get_future();
-  task_.bg_thread_ = std::thread([&](std::promise<common::Status> result_promise) {
-    common::Status s = Run(run_options, io_binding.GetInputNames(), io_binding.GetInputs(), io_binding.GetOutputNames(),
-                           &io_binding.GetOutputs(), &io_binding.GetOutputsDeviceInfo());
+    common::Status status = Run(run_options, io_binding.GetInputNames(), io_binding.GetInputs(), io_binding.GetOutputNames(),
+                                &io_binding.GetOutputs(), &io_binding.GetOutputsDeviceInfo());
 
-    result_promise.set_value(s);
+    onnxruntime::contrib::OrtTasks::GetInstance().SetStatus(status);
 
     // signal main thread for background thread completion
-    const int64_t main_thread_event_id = 0;
-    onnxruntime::contrib::OrtEventPool::GetInstance().SignalEvent(main_thread_event_id);
+    if (onnxruntime::contrib::OrtTasks::GetInstance().ForwardOutputsIsValid()) {
```

Comment:
added comments

    // If forward outputs still hasn't been consumed at this point, i.e. forward function hasn't complete itself
    // this indicates that Run() call returned before hitting YieldOp, due to hitting some exception during the forward subgraph execution
    // In this case, we need to wake up the foreground thread and pass along the failed status. 
    // Otherwise, foreground thread will be stuck waiting for forward_outputs.